File: ios/telobike/telobike/TBFavorites.m
//
//  Favorites.m
//  telobike
//
//  Created by ELAD BEN-ISRAEL on 11/13/11.
//  Copyright (c) 2011 __MyCompanyName__. All rights reserved.
//

#import "TBFavorites.h"
//#import "Analytics.h"

@implementation TBFavorites

+ (TBFavorites*)instance {
    static TBFavorites* i = nil;
    if (!i) {
        i = [[TBFavorites alloc] init];
    }
    return i;
}

- (NSString*)defaultsKeyForStationID:(NSString*)stationID
{
    return [NSString stringWithFormat:@"favorite.%@", stationID];
}

- (BOOL)isFavoriteStationID:(NSString*)stationID
{
    [[NSUserDefaults standardUserDefaults] synchronize];
    return [[NSUserDefaults standardUserDefaults] boolForKey:[self defaultsKeyForStationID:stationID]];
}

- (void)setStationID:(NSString*)stationID favorite:(BOOL)isFavorite
{
//    if (isFavorite) {
//        [[Analytics shared] eventAddFavorite:stationID];
//    }
//    else {
//        [[Analytics shared] eventRemoveFavorite:stationID];
//    }
    
    [[NSUserDefaults standardUserDefaults] setBool:isFavorite forKey:[self defaultsKeyForStationID:stationID]];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

@end

@implementation TBStation (Favorite)

- (BOOL)isFavorite {
    return [[TBFavorites instance] isFavoriteStationID:self.sid];
}

- (void)setFavorite:(BOOL)favorite {
    [[TBFavorites instance] setStationID:self.sid favorite:favorite];
}

@end

File: ios/telobike/telobike/TBFavorites.swift
//
//  TBFavorites.swift
//  telobike
//
//  Created by Elad Ben-Israel on 9/19/14.
//  Copyright (c) 2014 Elad Ben-Israel. All rights reserved.
//

import Foundation

class TBFavorites: NSObject {
    private var favorites = [String:Bool]()
    private let defaultsKey = "favorites"
    
    class var instance: TBFavorites {
        struct Singleton { static let instance = TBFavorites() }
        return Singleton.instance
    }
    
    override init() {
        if let dict = NSUserDefaults.standardUserDefaults().dictionaryForKey(defaultsKey) {
            self.favorites = dict as [String:Bool]
        }
    }
    
    private func defaultsKeyForStationID(stationID: String) -> String {
        return "favorite.\(stationID)"
    }

    func isFavoriteStationID(stationID: String) -> Bool {
        return self.favorites[stationID] ?? false
    }
    
    func setStationID(stationID: String, favorite isFavorite: Bool) {
        self.favorites[stationID] = isFavorite
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0)) {
            NSUserDefaults.standardUserDefaults().setObject(self.favorites, forKey: self.defaultsKey)
            NSUserDefaults.standardUserDefaults().synchronize()
        }
    }
}

extension TBStation {
    var favorite: Bool {
        get { return TBFavorites.instance.isFavoriteStationID(self.sid) }
        set { TBFavorites.instance.setStationID(self.sid, favorite: newValue) }
    }
    
    var isFavorite: Bool {
        return self.favorite
    }
}